File: src/dependencies/crazyflie_hardware/src/crtp_cpp/src/crtp_cpp/logic/link_layer_logic.cpp
#include "crtp_cpp/logic/link_layer_logic.hpp"

LinkLayerLogic::LinkLayerLogic(CrtpLink* crtp_link)
    : Logic(crtp_link),
      packer(LinkLayerPacker()) {}

void LinkLayerLogic::send_nullpacket() {
    CrtpRequest request;
    request.packet = packer.nullpacket();
    link->send_packet_no_response(request);
}

void LinkLayerLogic::platform_power_down() {
    CrtpRequest request;
    request.packet = packer.platform_power_down();
    link->send_packet_no_response(request);
}

void LinkLayerLogic::reboot_to_bootloader() {
    CrtpRequest request;
    request.packet = packer.reset_init();
    link->send_packet_no_response(request);

    request.packet = packer.reset_to_bootloader();
    link->send_packet_no_response(request);
}

void LinkLayerLogic::reboot_to_firmware() {
    CrtpRequest request;
    request.packet = packer.reset_init();
    link->send_packet_no_response(request);

    request.packet = packer.reset_to_firmware();
    link->send_packet_no_response(request);
}

File: src/dependencies/crazyflie_hardware/src/crtp/crtp/logic/link_layer_logic.py
from .logic import Logic
from crtp.crtp_link import CrtpLink
from crtp.packers.link_layer_packer import LinkLayerPacker
from crtp.packers.crtp_packer import CrtpPacker

from typing import Callable


class LinkLayerLogic(Logic):

    def __init__(
        self, crtp_packer_factory: Callable[[int], CrtpPacker], crtp_link: CrtpLink
    ):
        super().__init__(crtp_link)
        self.packer = LinkLayerPacker(crtp_packer_factory)

    def send_nullpacket(self):
        packet = self.packer.nullpacket()
        self.link.send_packet_no_response(packet)

    def platform_power_down(self):
        packet = self.packer.platform_power_down()
        self.link.send_packet_no_response(packet)

    def reboot_to_bootloader(self):
        packet = self.packer.reset_init()
        self.link.send_packet_no_response(packet)

        packet = self.packer.reset_to_bootloader()
        self.link.send_packet_no_response(packet)

    def reboot_to_firmware(self):
        packet = self.packer.reset_init()
        self.link.send_packet_no_response(packet)

        packet = self.packer.reset_to_firmware()
        self.link.send_packet_no_response(packet)
